src/ol/expr/expression.js

```diff
@@ -256,65 +268,101 @@ export function newParsingContext() {
 
 /**
  * @param {EncodedExpression} encoded The encoded expression.
+ * @param {number} expectedType The expected type.
  * @param {ParsingContext} context The parsing context.
- * @param {number} [typeHint] Optional type hint
  * @return {Expression} The parsed expression result.
  */
-export function parse(encoded, context, typeHint) {
+export function parse(encoded, expectedType, context) {
   switch (typeof encoded) {
     case 'boolean': {
+      if (isType(expectedType, StringType)) {
+        return new LiteralExpression(StringType, encoded ? 'true' : false);
+      }
+      if (!includesType(expectedType, BooleanType)) {
+        throw new Error(
+          `got a boolean, but expected ${typeName(expectedType)}`,
+        );
+      }
       return new LiteralExpression(BooleanType, encoded);
     }
     case 'number': {
-      return new LiteralExpression(
-        typeHint === SizeType ? SizeType : NumberType,
-        encoded,
-      );
+      if (isType(expectedType, SizeType)) {
+        return new LiteralExpression(SizeType, toSize(encoded));
+      }
+      if (isType(expectedType, BooleanType)) {
+        return new LiteralExpression(BooleanType, !!encoded);
+      }
+      if (isType(expectedType, StringType)) {
+        return new LiteralExpression(StringType, encoded.toString());
+      }
+      if (!includesType(expectedType, NumberType)) {
+        throw new Error(`got a number, but expected ${typeName(expectedType)}`);
+      }
+      return new LiteralExpression(NumberType, encoded);
```

Comment:
When parsing literals in expressions, we cast values when appropriate to the expected type.